File: apps/web/src/lib/admin/reported-content.ts
import { authorizedFetch } from '../api/client';
import { errorFrom } from '../api/problem';

/**
 * What a report is about — issue #399.
 *
 * <h2>The read the moderation queues were deciding without</h2>
 *
 * <p>`/admin/moderation/{id}` is the deepest view of a complaint, and for a report about a
 * comment it rendered: the reason, the reporter's claim, an identifier for the reporter, and
 * two irreversible buttons. Counted in the DOM, zero links. Not the comment, not who wrote
 * it, not which campaign it is on, and no route to any of the three.
 *
 * <p>A moderator asked to uphold or dismiss something they cannot read either guesses or
 * leaves the queue to find the content another way — and the cheap guess is always to
 * dismiss, because upholding a report you cannot read is the riskier of the two. That is
 * how a moderation queue stops working while continuing to look like it works.
 *
 * <h2>A third read, and it does not block the other two</h2>
 *
 * <p>The detail screen already loads the report and its audit trail independently, so a
 * trail that fails costs a line of text rather than the screen. This is the third read in
 * that arrangement and degrades the same way: a moderator who cannot load the evidence is
 * told the evidence is missing, and still sees the report.
 *
 * <p>It is not folded into the report's own response because that response is also every row
 * of the queue — assembling the content, its author and its campaign twenty-five times a page
 * is twenty-five lookups nobody reads.
 */

/**
 * What the platform can say about the reported thing.
 *
 * <p>Four values rather than a success flag, because they lead to four different decisions:
 *
 * - `PRESENT` — it is there and it is on the platform.
 * - `REMOVED` — it is there and it has been taken down. **The text still comes back**: the
 *   service keeps a removed comment's row for exactly this, and a moderator told only
 *   "removed" cannot tell an upheld report from a dismissed one. What the screen adds is
 *   that somebody has already handled it, so it is not handled twice.
 * - `GONE` — the identifier names nothing any more. Distinct from `REMOVED` on purpose.
 * - `ADDRESSED_DIRECTLY` — the target is a campaign or an account, which the console reaches
 *   directly: a campaign has a staff preview that renders it in any state, and an account has
 *   a public profile. There is no separate blob of text to inline, and a fragment of a page
 *   next to a link to that page is worse than the link alone.
 */
export type ReportedContentState = 'PRESENT' | 'REMOVED' | 'GONE' | 'ADDRESSED_DIRECTLY';

/** The campaign the content sits on, with both halves of its public path. */
export interface ReportedCampaign {
  id: string;
  title: string;
  slug?: string | null;
  creatorSlug?: string | null;
}

/**
 * The evidence, as the console renders it.
 *
 * <p>Every optional field is `?: T | null` because the service serialises with
 * `default-property-inclusion: non_null` — an absent title is absent from the JSON rather
 * than present and null.
 */
export interface ReportedContent {
  /** `COMMENT`, `PROJECT_UPDATE`, `PROJECT` or `USER` — the report's own target kind. */
  targetType: string;
  state: ReportedContentState;
  /** An update's headline. Absent for a comment, which has none. */
  title?: string | null;
  /**
   * What was written, verbatim.
   *
   * <p><strong>Untrusted.</strong> It is text one member of the public wrote about another,
   * arriving on a screen operated by staff, and it is rendered as text and never as markup.
   */
  body?: string | null;
  authorId?: string | null;
  /** An update's per-campaign sequence — "update 4", which is what people call it. */
  number?: number | null;
  project?: ReportedCampaign | null;
  /** When the content was written, which is not when the report was filed. */
  createdAt?: string | null;
}

/**
 * The evidence behind one report.
 *
 * @throws ApiError 404 when there is no such report. **Not** when the content has gone —
 *     that is a 200 carrying `GONE`, because "there is no such report" and "the comment it
 *     was about has been purged" send a moderator to two different places
 */
export async function readReportedContent(
  reportId: string,
  signal?: AbortSignal,
): Promise<ReportedContent> {
  const response = await authorizedFetch(
    `/v1/admin/moderation/reports/${encodeURIComponent(reportId)}/content`,
    { signal },
  );
  if (!response.ok) throw await errorFrom(response);

  return (await response.json()) as ReportedContent;
}

File: apps/api/src/main/java/az/ideanest/moderation/application/ReportedContent.java
package az.ideanest.moderation.application;

import az.ideanest.community.application.ModeratedContent;
import az.ideanest.community.application.ModeratedContent.ModeratedComment;
import az.ideanest.community.application.ModeratedContent.ModeratedUpdate;
import az.ideanest.moderation.domain.ContentReport;
import az.ideanest.moderation.domain.ReportTargetType;
import az.ideanest.moderation.infrastructure.ContentReportRepository;
import az.ideanest.shared.access.PlatformStaff;
import az.ideanest.shared.project.ProjectSummaries;
import az.ideanest.shared.project.ProjectSummary;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * The evidence behind one report — #399.
 *
 * <h2>A third read, not a wider one</h2>
 *
 * <p>{@code ReportModerationService.report} serves the complaint and its decision history,
 * and this serves what the complaint is about. Two services rather than one field added to
 * {@link QueuedReport}, because that record is also every row of the queue: assembling the
 * comment, its author and its campaign for twenty-five rows is twenty-five lookups nobody
 * reads, on the one screen in the console that has to stay fast enough to work from.
 *
 * <p>The detail page already reads the report and its audit trail independently, so that a
 * trail which fails to load costs a line of text rather than the screen. This is the third
 * read in that arrangement and it degrades the same way — a moderator who cannot load the
 * evidence should be told the evidence is missing, and should still see the report.
 *
 * <h2>The refusal is here, and it is the queue's own</h2>
 *
 * <p>{@code requireStaff}, matching {@code ReportModerationService.report} exactly. This
 * endpoint says nothing about a report that a moderator holding the report cannot already
 * see, so a stricter check would refuse somebody the evidence for a decision they are
 * entitled to take. Checked before the report is loaded, so a caller who is not staff
 * learns nothing about which report identifiers are real.
 *
 * <h2>Every kind of target is answered, including the two with nothing to inline</h2>
 *
 * <p>A campaign and an account are reached directly by the console — the staff preview at
 * {@code /admin/campaigns/{id}} renders a campaign in any state, and a profile has a public
 * page — so this returns {@link ReportedItem.State#ADDRESSED_DIRECTLY} for them rather than
 * a 404 or an empty body. One shape for four kinds is what lets the screen render the
 * answer without knowing which kind it asked about.
 */
@Service
public class ReportedContent {

    private final ContentReportRepository reports;
    private final ModeratedContent content;
    private final ProjectSummaries projects;
    private final PlatformStaff staff;

    public ReportedContent(
            ContentReportRepository reports,
            ModeratedContent content,
            ProjectSummaries projects,
            PlatformStaff staff) {

        this.reports = reports;
        this.content = content;
        this.projects = projects;
        this.staff = staff;
    }

    /**
     * What this report is about.
     *
     * @param moderatorId whoever is signed in, from the access token's subject
     * @throws ReportNotFoundException when there is no such report. Never for content that
     *     has gone: that is {@link ReportedItem.State#GONE}, and the difference matters —
     *     "there is no such report" and "the comment it was about has been purged" send a
     *     moderator to two different places
     */
    @Transactional(readOnly = true)
    public ReportedItem of(UUID reportId, UUID moderatorId) {
        staff.requireStaff(moderatorId);

        ContentReport report = reports.findById(reportId).orElseThrow(() -> new ReportNotFoundException(reportId));

        return switch (report.getTargetType()) {
            case COMMENT -> comment(report.getTargetId());
            case PROJECT_UPDATE -> update(report.getTargetId());
            // The campaign the report names, so the console can link to its staff preview.
            case PROJECT -> ReportedItem.addressedDirectly(
                    report.getTargetType(), projects.summaryOf(report.getTargetId()).orElse(null));
            // An account belongs to no campaign, and the console names it through the
            // directory #402 built rather than through a field on this response.
            case USER -> ReportedItem.addressedDirectly(report.getTargetType(), null);
        };
    }

    private ReportedItem comment(UUID commentId) {
        ModeratedComment found = content.comment(commentId).orElse(null);
        if (found == null) {
            return ReportedItem.gone(ReportTargetType.COMMENT);
        }
        return ReportedItem.comment(
                found.removed() ? ReportedItem.State.REMOVED : ReportedItem.State.PRESENT,
                found.body(),
                found.authorId(),
                campaign(found.projectId()),
                found.createdAt());
    }

    private ReportedItem update(UUID updateId) {
        ModeratedUpdate found = content.update(updateId).orElse(null);
        if (found == null) {
            return ReportedItem.gone(ReportTargetType.PROJECT_UPDATE);
        }
        return ReportedItem.update(
                found.title(),
                found.body(),
                found.authorId(),
                found.number(),
                campaign(found.projectId()),
                found.publishedAt());
    }

    /**
     * The campaign the content is on, or null.
     *
     * <p>Null rather than a failure, which is {@link ProjectSummaries}' own contract and the
     * right one here: a comment whose campaign has been hard deleted is still a comment
     * somebody complained about, and losing the evidence over a missing link would be the
     * console failing at exactly the moment it is most needed.
     */
    private ProjectSummary campaign(UUID projectId) {
        return projects.summaryOf(projectId).orElse(null);
    }
}
